Install & run PHP-CS-Fixer

DominicLuidold opened the pull request:
| Q             | A  |
|---------------|----|
| Bug fix?      | no |
| New feature?  | no |
| Deprecations? | no |
| Issues        | /  |

As there are currently efforts to introduce PHPStan (see #2249), I thought of also introducing PHP-CS-Fixer as a dev dependency (to additionally improve the code quality) and using/upgrading the existing config file. 

**Already done:**
- Introduce [`PHP-CS-Fixer`](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer) as a dev dependency
- Upgrade the existing config
- Run PHP-CS-Fixer on the project files

**Possible further steps as part of this PR that need to be discussed:**
- Replace StyleCI with a PHP-CS-Fixer job as part of the GitHub Actions `CI`/`continuous-integration` workflow (to have a single source of truth regarding code style)?
  - Run PHP-CS-Fixer on same test matrix as PHPUnit[^1] or just on latest supported PHP version (e.g. 8.3)?

[^1]: This approach will not work on PHP 7.4 and PHP 8.0 as some files already use newer syntax leading to failures in these jobs.

DominicLuidold commented on .php-cs-fixer.dist.php:
```diff
@@ -1,16 +1,15 @@
+return (new PhpCsFixer\Config())
     ->setRules([
         '@Symfony' => true,
-        'ordered_imports' => true,
-        'phpdoc_order' => true,
```
These two rules are already part of the `@Symfony` rule-set

DominicLuidold commented on .php-cs-fixer.dist.php:
```diff
@@ -21,6 +20,6 @@
 file that was distributed with this source code.
 HEADER
         ],
+        'nullable_type_declaration_for_default_null_value' => false,
```
This is required to maintain BC - possibly a rule that could be removed (meaning following the `@Symfony` rule-set which has set this rule to `true`) as part of #2215

DominicLuidold replied:
Actually, this wouldn't even provoke a BC break, see https://3v4l.org/UkVXL.

So this could definitely be removed already to align with the `@Symfony` rule-set and to be prepared for the next PHP minor version which will deprecate the implicit nullable parameters (see https://wiki.php.net/rfc/deprecate-implicitly-nullable-types).

DjordyKoert replied:
Yeah I agree that we should remove this 👍

DjordyKoert commented:
> **Possible further steps as part of this PR that need to be discussed:**
> 
> * Replace StyleCI with a PHP-CS-Fixer job as part of the GitHub Actions `CI`/`continuous-integration` workflow (to have a single source of truth regarding code style)?
>   

I also think that we should remove StyleCI in favor of a PHP-CS-Fixer job. I would have to message/email @GuilhemN though asking if he can disable StyleCI for this project before merging this though. As far as I can see there is no easy way to disable it using a simple config change in `.styleci.yml`

>   * Run PHP-CS-Fixer on same test matrix as PHPUnit[1](#user-content-fn-1-6d0da86d1317f5d8de24e36bde6fba6e) or just on latest supported PHP version (e.g. 8.3)?

I think running it on the latest PHP version (8.3) will suffice.

DominicLuidold commented:
I've refactored the current CI workflow a bit and introduced a [composite action](https://docs.github.com/en/actions/creating-actions/creating-a-composite-action) that allows to re-use the steps for setting up and installing the Composer dependencies across multiple jobs.

This change should make it quite easy to add a PHPStan CI job in #2249 as well :smile:

DjordyKoert commented:
Update: I have sent an email to @GuilhemN asking to disable StyleCI for this repo

DjordyKoert commented:
StyleCI has now been disabled, I will apply my own suggestion and merge it asap

DjordyKoert commented:
Thank you very much @DominicLuidold for your work ❤️
